PAC and Midrow code handling for EIA-608 captions

RikHeijdens opened the pull request:
This is an effort to implement preamble control codes and mid row style changes for EIA-608 captions.

Currently the implementation does not completely follow the EIA-608 spec, but should provide reasonable closed caption positioning and styling for most cases.

You can test this by Selecting the `Apple master playlist` stream under HLS.

Commit:
Implemented PAC and Midrow code handling for EIA-608 captions

RikHeijdens commented on library/src/main/java/com/google/android/exoplayer2/text/eia608/Eia608Decoder.java:
```diff
@@ -414,32 +501,197 @@ private void handleMiscCode(byte cc2) {
+      // the new line receives it's own PAC which we ignore currently.
+      // As a result of that the new line will be positioned directly below the
+      // previous line.
+      maybeAppendNewline();
```
Ideally, we should create a new Cue here, but I'm not sure how we can display Cues within different Eia608Subtitle objects on the screen at the same time. Maybe we shouldn't try to instantiate new Eia608Subtitle objects in `dequeueOutputBuffer()` all the time, and instead stream the cues through a single instance of Eia608Subtitle?

Commit:
Discard spans after seeking

Discard any spans that should not be applied because they don't belong
to what's in the current CC buffer.

ojw28 commented:
@cdrolle FYI

RikHeijdens commented on library/src/main/java/com/google/android/exoplayer2/text/eia608/Eia608Decoder.java:
```diff
@@ -414,32 +501,197 @@ private void handleMiscCode(byte cc2) {
+    }
+  }
+
+  private void handleMidrowCode(byte cc1, byte cc2) {
```
I don't have a stream that uses mid row codes to change the foreground/background color to anything but white/black, so I'm not sure if this actually works.

zsmatyas replied:
Check the implementation here:
https://github.com/google/ExoPlayer/pull/1312/files#diff-e9fd670327a94df1ab96714b095f4985R397

Beware the order of the priorities.

RikHeijdens replied:
This should take care of the order of priorities: https://github.com/google/ExoPlayer/pull/1773/commits/7cd94819c7a4a0222022cbfd49a1d77f30aa5812#diff-7fa75103163e987aeb1657e153a4604fR596

Do you by any chance have a stream that uses midrow codes to change styles? Apple's HLS sample doesn't seem to do it.

Commit:
Fixed comment

Alpha is stored in the first byte of an integer

zsmatyas commented on library/src/main/java/com/google/android/exoplayer2/text/eia608/Eia608Decoder.java:
```diff
@@ -414,32 +501,201 @@ private void handleMiscCode(byte cc2) {
+    // We're ignoring any tab offsets that do not occur at the beginning of a new cue.
+    // This is not conform the spec, but works in most cases.
+    if (captionStringBuilder.length() == 0) {
+      switch (cc2) {
```
This entire block can be simplified to a single line: 
tabOffset = cc2 - 0x20;

zsmatyas commented on library/src/main/java/com/google/android/exoplayer2/text/eia608/Eia608Decoder.java:
```diff
@@ -86,6 +99,12 @@
 
+  private static final byte CTRL_TAB_OFFSET_CHAN_1 = 0x17;
+  private static final byte CTRL_TAB_OFFSET_CHAN_2 = 0x1F;
+  private static final byte CTRL_TAB_OFFSET_1 = 0x21;
```
These can be removed.

zsmatyas commented on library/src/main/java/com/google/android/exoplayer2/text/eia608/Eia608Decoder.java:
```diff
@@ -159,23 +178,79 @@
 
+  // Maps EIA-608 PAC row numbers to WebVTT cue line settings.
+  // Adapted from: https://dvcs.w3.org/hg/text-tracks/raw-file/default/608toVTT/608toVTT.html#x1-preamble-address-code-pac
+  private static final float[] CUE_LINE_MAP = new float[] {
```
If you want to use a map and the standard has the row values in specific, but not linear order like: (1, 3, 2).

And than you reorder everything with a switch case, like you map the value 0x1 to the index 4:

case 0x1:  // Row 5 or 6
   cueLine = CUE_LINE_MAP[4];

Then you could put those values into the map in the same order as the standard specifies, and than you can remove the switch and directly index the map:
cueLine[ 0x1 ];

See previous implementation:
https://github.com/google/ExoPlayer/pull/1312/files#diff-babdf8243982d21611d6cb14f7df1bf9R104

RikHeijdens replied:
Thanks - I didn't know about your implementation.
I'll address all the comments tomorrow.

zsmatyas commented on library/src/main/java/com/google/android/exoplayer2/text/eia608/Eia608Decoder.java:
```diff
@@ -414,32 +501,201 @@ private void handleMiscCode(byte cc2) {
+  }
+
+  private int getSpanStartIndex() {
+    return captionStringBuilder.length() > 0 ? captionStringBuilder.length() - 1 : 0;
```
Will this call length() twice or is the compiler clever enough to call it only once?

RikHeijdens replied:
I think it should be clever enough, but I didn't check the byte code. Might be worth a little investigation.

zsmatyas commented on library/src/main/java/com/google/android/exoplayer2/text/eia608/Eia608Decoder.java:
```diff
@@ -414,32 +501,201 @@ private void handleMiscCode(byte cc2) {
   private void backspace() {
     if (captionStringBuilder.length() > 0) {
-      captionStringBuilder.setLength(captionStringBuilder.length() - 1);
+      captionStringBuilder.replace(captionStringBuilder.length() - 1, captionStringBuilder.length(), "");
```
Probably more efficient:
https://docs.oracle.com/javase/7/docs/api/java/lang/StringBuilder.html#delete(int,%20int)

RikHeijdens replied:
According to the [documentation](https://developer.android.com/reference/android/text/SpannableStringBuilder.html#delete%28int, int%29) delete is just a convenience method, but I agree, it would probably fit better here.

zsmatyas commented on library/src/main/java/com/google/android/exoplayer2/text/eia608/Eia608Decoder.java:
```diff
@@ -463,19 +719,19 @@ private String getDisplayCaption() {
     int searchBackwardFromIndex = endIndex;
     for (int i = 0; i < captionRowCount && searchBackwardFromIndex != -1; i++) {
-      searchBackwardFromIndex = captionStringBuilder.lastIndexOf("\n", searchBackwardFromIndex - 1);
+      searchBackwardFromIndex = captionStringBuilder.toString().lastIndexOf("\n", searchBackwardFromIndex - 1);
```
This entire logic can be removed as soon as every line is implemented with a separate Cue object, but until then:
what is the problem with the lastIndexOf() of StringBuilder? Why do you use the String one?
https://docs.oracle.com/javase/7/docs/api/java/lang/StringBuilder.html#lastIndexOf(java.lang.String,%20int)

RikHeijdens replied:
[SpannableStringBuilder](https://developer.android.com/reference/android/text/SpannableStringBuilder.html#delete%28int, int%29) does not offer the lastIndexOf method.

And yes, most likely it can be removed after we figured out: https://github.com/google/ExoPlayer/pull/1773#discussion_r76045863

zsmatyas replied:
You are right, sorry. I incorrectly assumed SpannableStringBuilder being a subclass of StringBuilder

Commit:
Improved line and tab offset mapping

Commit:
Took care of caption styling priorities

zsmatyas commented on library/src/main/java/com/google/android/exoplayer2/text/eia608/Eia608Decoder.java:
```diff
@@ -159,23 +175,72 @@
+  private static final int TRANSPARENCY_MASK = 0x80FFFFFF;
+
+  private static final int STYLE_ITALIC = Typeface.ITALIC;
+  private static final float DEFAULT_CUE_LINE = CUE_LINE_MAP[4]; // Row 11
```
the 4th element does not seem to be line 11. Is the comment matching the code?

RikHeijdens replied:
Thanks, good spot!

It will be fixed in the next commit, I'm currently working on displaying every line as a separate cue.

Commit:
Caption lines as separate cues

RikHeijdens commented:
Since https://github.com/google/ExoPlayer/pull/1773/commits/71d83f3e8478f05271845eb36a0a3f1bfec98808 every caption line is now a separate cue and being streamed through a single instance of `Eia608Subtitle`, at this point of time that causes a couple of issues:
- Cue background boxes overlap
- Paint-on/Roll-up captions are not being drawn correctly

Furthermore, I'm not happy about horizontal positioning. The EIA-608 standard assumes a maximum of 32 characters per line, because of that horizontal positioning is off in general. I'm not sure what would be the best approach to handle it. I checked AV Player's behavior and they seem to align cues with less than 1 empty row between them. Should we mimic that behavior?

zsmatyas commented:
Good job, you are almost there. :) If you can wait a few days, I might be able to push my V1 changes ported to V2, I just lowered priority working on these as I did not see too much chance that my pull request gets accepted.

To address the issues:
- Overlap is caused because with the correct positioning and default sizes there must be an overlap. (You cannot fit 15 thicker lines into the same area that was using narrower lines). I solved this by slicing the rendering into 2 phases: backgrounds first then the texts. It is just a few lines of code.
-  Paint on cannot work nicely with fonts that are not unified width. Replacing any character with one of different with will re-align the entire line, the text continuously moves, you cannot read it. 
- For Roll up cations the line count, new line and the "show the text" commands must be implemented according to the spec. Remove hacks like maybeAppendNewLine(). These were probably included in my commits for the V1 implementation. The output of my version (colors/positioning/timing) matched the other players for android that handles 608.

Bad news: 708 will complicate things much more :)

RikHeijdens commented:
@zsmatyas 

Roll-up/Paint-on captions should follow the spec pretty close now, but for some reason the SubtitlePainter does not paint the entire CueText. For example with the following [stream](http://playertest.longtailvideo.com/adaptive/cpcweb/cpcdemo.m3u8) it only paints the first two characters of the paint-on cue, when the SubtitlePainters receive the entire Cue. I'm not sure why that is happening, as it should recalculate the layout.

Commit:
Fixed foreground color midrow codes being handled as background colors

Commit:
Return a copy of the SpannableStringBuilder when building Cues

In order to force the SubtitlePainter to redraw cues

RikHeijdens commented:
OK, disregard my previous comment. That should be fixed now :)

Commit:
Split Cue Rendering up into 3 phases

Split Cue Rendering up into a layout, background drawing and foreground
drawing phase so that issues with overlapping 608 captions are being prevented.

Commit:
Removed unnecessary member variable

yichengliu commented on library/src/main/java/com/google/android/exoplayer2/text/eia608/Eia608CueBuilder.java:
```diff
@@ -0,0 +1,200 @@
+   * @return true if rolling was possible.
+   */
+  public boolean rollUp() {
+    if (row < 1) {
```
It seems the condition here should be `if (row <= 1)`. Otherwise when `row` equals `1`, `setRow(0)` will be called and there will be an exception.

yichengliu commented on library/src/main/java/com/google/android/exoplayer2/text/eia608/Eia608Decoder.java:
```diff
@@ -414,68 +453,102 @@ private void handleMiscCode(byte cc2) {
+          for (Eia608CueBuilder cue : cues) {
+            // Roll up all the other rows.
+            if (!cue.rollUp()) {
+              cues.remove(cue);
```
This code can cause `ConcurrentModificationException` when `rollUp()` returns false.

RikHeijdens replied:
Thanks @YichengLiu I've addressed both this issue and your other comment.

Commit:
Fixed an off by one error and fixed iterating through the cues

Commit:
Corrected vertical positioning

RikHeijdens commented:
@ojw28 @cdrolle I noticed that recently some changes where made to the text.eia608/cea packages - is this something that eventually will be merged or are you working on your own implementation?

ojw28 commented:
I'm aware there are some issues with the merged result of this. We'll get those fixed up shortly!

ojw28 commented:
Thanks for the contribution btw!
